rfcs/0000-terraform-workflow-improvements.md

```diff
@@ -0,0 +1,826 @@
+  - Option 2: Allow overriding image tag in module with [`.tfvars` files][terraform-tfvars].
+
+### Isolated Testing Environment
+
+As noted in the [guide-level explanation][guide-module-testing], testing will happen in the separate `mbta-ctd-test` AWS account. This account will be accessed using "assume role" permissions that full access to manage required AWS resources. The assumed role won't have full administrator access, but we'll manage the list of permitted services via IAM policies.
+
+## 3. Standardizing on Infrastructure Component Modules
+[reference-standardization]: #3-standardizing-on-infrastructure-component-modules
+
+As noted in the [guide-level explanation][guide-standardization], this is currently the most mature of the proposed improvements, in that the `ctd-ecs-app` and `ctd-rds-db` modules already exist and are seeing use. Aside from increasing adoption and carrying out the [testing workflow][guide-module-testing] and [module versioning][guide-module-versioning] recommendations, other improvements that will be beneficial include:
+
+- Handling all use cases for each module, using vars to enable/disable features that aren't likely to be used everywhere
+- Building in support for maintenance modes to minimize temporary configuration changes
+- Standardizing on the use of the `context` variable
+
+## 4. Module Versioning/Publishing
+[reference-module-versioning]: #4-module-versioningpublishing
+
+As noted in the [guide-level explanation][guide-module-versioning], module versioning and publishing is something we will set up for all infrastructure component modules, but which should be considered optional for application modules. See also the [corresponding subsection under rationale and alternatives][alternatives-module-versioning].
+
+### How To Publish A Module
+
+To set up a module for versioning and publishing, we'll need to:
+
+- Move the module's configuration to a separate repo, for a few reasons:
+  - It simplifies the workflow for tagging/publishing releases
+  - It's required if publishing to the public Terraform registry
+- Ensure that the repo name follows the convention `terraform-aws-<module-name>`, which is a requirement for both [Terraform's public registry][terraform-module-publishing] and [Scalr's private registry][scalr-private-registry]
+- Set up a tagging/release workflow in GitHub Actions that publishes the module to Scalr's private module registry or the public Terraform registry (ideally as a reusable workflow)
+- Configure any references in root modules to call the module from the repo or registry instead of a relative path to the module
+
+### Application Modules In Application Repos?
+
+For application modules, the question has been raised as to whether it might make sense to move the module configuration into the application repo so it lives alongside the codebase. The stated advantages to this approach are as follows:
+
+- All(?) of the application's configuration would live within the same repo
```

Comment:
An advantage I'd been thinking of to storing the config alongside the code is that deploying the code could also apply the config, so that the code would always see, for example, the environment variables it was expecting. I still think this would be good, but it would introduce some new challenges, and it seems like it'd require a complete overhaul of the entire structure of our Terraform setup to get to a point where that's even possible.

If the config and the code are deployed independently, then they should be versioned independently, so I agree that it makes more sense for them to be stored independently.